designs/2018-relative-package-loading/README.md

```diff
@@ -0,0 +1,278 @@
+
+Another way to state the problem is that ESLint's mechanism for naming rules in a config file (`plugin-name/rule-name`) is fundamentally unable to disambiguate two plugins that have the same name. Currently, there are no naming conflicts because all plugins are loaded from the location of the `eslint` package, so plugins effectively live in a global namespace. If plugins could be loaded as dependencies of shareable configs, as implied by this proposal, then naming conflicts would start to become a problem.
+
+Note: The name ambiguity problem applies to all named resources provided by a plugin. This includes rules and environments, and will also include other resources such as processors if [RFC #3](https://github.com/eslint/rfcs/pull/3) is approved. For brevity, the rest of this RFC only refers to rule naming, but the same solution will also be used to name other plugin-provided resources.
+
+### Solving the duplicate-plugin problem
+
+Since the problem stems from a limitation of ESLint's naming scheme, the most straightforward solution to the problem would involve changing the way that rules are named in a config file. Specifically, rules will now be addressed with hierarchical naming for rules, where rules can be configured with something that looks like a path. (In other words, a user could refer to a rule like `foo::react/some-rule` for the version of `eslint-plugin-react` used by `eslint-config-foo`, and this would be a different rule than `bar::react/some-rule`, which would refer to the version of `eslint-plugin-react` used by `eslint-config-bar`.) To preserve compatibility, the existing naming scheme `react/some-rule` would continue to work in cases where the reference is unambiguous.
+
+The following sections describe the specifics of how hierarchical naming works.
+
+#### Config trees
+
+When describing how rule name resolution works in this proposal, it's useful to think of a "config tree" representing the dependencies between shareable configs and plugins. The root node is the end user's config, and each node has a set of named children representing the shareable configs that it extends and plugins that it depends on. Here's an example tree:
+
+![](./config-tree-example.png)
+
+In this example, the end user's config extends `eslint-config-foo` and `eslint-config-bar`. `eslint-config-bar` extends `eslint-config-baz`. `eslint-config-foo` and `eslint-config-baz` both depend on versions of `eslint-plugin-react` (perhaps different versions, although this doesn't matter as far as resolution is concerned). `eslint-config-baz` also depends on `eslint-plugin-import`.
+
+#### Details of hierarchical rule name resolution
+
+* Each reference to a plugin rule in a config consists of three parts: a *config scope* (i.e. a list of configs), a plugin name, and a rule name.
```

Comment:
There are two cases here:

* If `plugin:react/recommended` configures `eslint-plugin-react`'s own rules (e.g. `react/no-typos`, you can just refer to the rule as `react/no-typos`. This is because if a plugin config configures its own rules, the rule references refer to the plugin itself rather than the plugin's dependencies (as described in [this conditional](https://github.com/eslint/rfcs/blame/55659901143fb933c3baa411bf6f8071730172d9/designs/2018-relative-package-loading/README.md#L78)).
* If `plugin:react/recommended` itself depends on another plugin `baz`, and you want to configure `baz`'s rules (and there is also another version of `baz` present that could create ambiguities), then you could identify the rule with `"plugin:react/recommended::baz/rule-name"`. In other words, `"plugin:react/recommended"` would be the config name that appears in the scope. I've added an example of this to the "Examples of hierarchical rule name resolution" section.